### ISSUE TITLE
Fft fit with binding energy scale does not work 

### ISSUE THREAD

**Issue by SunC0der** [Julian-Hochhaus/lmfit-additional-models#6]
Hi, 
I am trying currently to fit spectra which are on the binding energy axis, in contrast to the spectra previously which where on the kinetic energy axis.  
Somehow changing the fft voigt peak center values (with  min / max values) and the peak parameter order does not produce a reasonable fit, see plot below. 

Which other parameter do I need to change to be able to fit the data on a binding energy scale ?
Does it have to do with the way I define my fft model ? 

## Plot:
![0-CTF3_21 31 csv](https://github.com/Julian-Hochhaus/lmfit-additional-models/assets/114517513/3a1cb185-ed2d-4b2c-ba88-b18397cc201e)

## Fit report:
[fit_report_0-CTF3_21&31.csv](https://github.com/Julian-Hochhaus/lmfit-additional-models/files/11781898/fit_report_0-CTF3_21.31.csv)

## Data file:
[0-CTF3_21&31.csv](https://github.com/Julian-Hochhaus/lmfit-additional-models/files/11781899/0-CTF3_21.31.csv)

## Code: 

```python

from IPython import display
import os
import sys
import re

sys.path.append(
    "/path/to/lmfit-additional-models"
)

from pathlib import Path

from models import ShirleyBG
from models import fft_convolve
import lmfit
from lmfit import Minimizer
from lmfit.model import save_modelresult
from lmfit.models import guess_from_peak
from lmfit.models import (
    VoigtModel,
    LorentzianModel,
    GaussianModel,
    LinearModel,
)
from lmfit.lineshapes import gaussian, lorentzian

import numpy as np
import pandas as pd
from scipy.special import wofz
from scipy.integrate import simpson as simps

import matplotlib
import matplotlib.pyplot as plt
from PIL import Image, ImageDraw, ImageFont
import time

# Define fft voigt model -----------------------------------------------------

def fft_voigt_2(x, amplitude=1.0, center=0.0, sigma=1.0, gamma=1):
    conv_temp = fft_convolve(
        lorentzian(x, amplitude=1, center=center, sigma=gamma),
        1
        / (np.sqrt(2 * np.pi) * sigma)
        * gaussian(x, amplitude=1, center=np.mean(x), sigma=sigma),
    )
    area = simps(conv_temp, x)
    return amplitude * conv_temp / area


class FFTVoigtModel(lmfit.model.Model):
    def __init__(
        self,
        independent_vars=["x"],
        prefix="",
        nan_policy="raise",
        *args,
        **kwargs,
    ):
        kwargs.update(
            {
                "prefix": prefix,
                "nan_policy": nan_policy,
                "independent_vars": independent_vars,
            }
        )
        super().__init__(fft_voigt_2, *args, **kwargs)
        self._set_paramhints_prefix()

    def _set_paramhints_prefix(self):
        self.set_param_hint("amplitude", value=25, min=0)
        self.set_param_hint("sigma", value=0.2, min=0)
        self.set_param_hint("gamma", value=0.02, min=0)
        self.set_param_hint("center", value=500)

        # Set expressions for FWHM and height for fit report
        fexpr = f"1.0692*{self.prefix}gamma + sqrt(0.8664*{self.prefix}gamma**2+5.545083*{self.prefix}sigma**2)"
        hexpr = f"({self.prefix}amplitude/(max(0, {self.prefix}sigma*sqrt(2*pi))))*wofz((1j*{self.prefix}gamma)/(max(0, {self.prefix}sigma*sqrt(2)))).real"

        # Add height and FWHM as parameters for fit report
        self.set_param_hint("fwhm", expr=fexpr)
        self.set_param_hint("height", expr=hexpr)

    def guess(self, data, x=None, **kwargs):
        if x is None:
            return
        lorentzian_pars = guess_from_peak(
            LorentzianModel(), data, x, negative=False
        )
        gaussian_pars = guess_from_peak(
            GaussianModel(), data, x, negative=False
        )
        params = self.make_params(
            amplitude=lorentzian_pars["amplitude"].value,
            sigma=gaussian_pars["sigma"].value,
            gamma=lorentzian_pars["sigma"].value,
            center=lorentzian_pars["center"].value,
        )
        return lmfit.models.update_param_vals(params, self.prefix, **kwargs)

# Define fitting function ----------------------------------------------------------------------------------------------------------------------------

def fit_four_fftvoigt_binding(
    file,
    vg1_center,
    vg1_amplitude,
    vg2_center,
    vg2_amplitude,
    vg3_center,
    vg3_amplitude,
    vg4_center,
    vg4_amplitude,
):
    """Function to fit four FFTvoigt functions with background."""
    # General adaption:
    matplotlib.interactive(
        False
    )  # prevents spyder from floding the Ipython terminal with <Figure size 432x288 with 0 Axes> during batch processing

    # Load and label data:
    file = Path(file)
    directory = file.resolve().parent

    fit_reports_directory = directory / "fit_reports"
    if fit_reports_directory.is_dir():
        pass
    else:
        fit_reports_directory.mkdir()

    fit_plots_directory = directory / "fit_plots"
    if fit_plots_directory.is_dir():
        pass
    else:
        fit_plots_directory.mkdir()

    data = pd.read_csv(
        str(file),
        sep=",",
        header=None,
    )

    data.columns = [
        "Binding Energy [eV]",
        "Kinetic Energy [eV]",
        "Intensity [a.u.]",
    ]
    x = data.loc[:, "Binding Energy [eV]"]
    y = data.loc[:, "Intensity [a.u.]"]

    # lmfit - Define models

    # shirley_model = ShirleyBG(prefix="shirley_", independent_vars=["y"])

    linear_model = LinearModel(prefix="lin_", independent_vars=["x"])
    background_model = linear_model  # + shirley_model

    Pb2_4f_52_peak = FFTVoigtModel(
        prefix="vg1_",
        independent_vars=["x"],  # nan_policy="omit"
    )
    Pb2_4f_72_peak = FFTVoigtModel(
        prefix="vg2_",
        independent_vars=["x"],  # nan_policy="omit"
    )
    Pb0_4f_52_peak = FFTVoigtModel(
        prefix="vg3_",
        independent_vars=["x"],  # nan_policy="omit"
    )
    Pb0_4f_72_peak = FFTVoigtModel(
        prefix="vg4_",
        independent_vars=["x"],  # nan_policy="omit"
    )

    combined_model = (
        Pb2_4f_72_peak
        + Pb2_4f_52_peak
        + Pb0_4f_72_peak
        + Pb0_4f_52_peak
        + background_model
    )  # changed '-' to '+', because your data contain background + data, model needs to reflect that, background is included in the model and fit and not removed before fitting

    # Define parameters

    params = combined_model.make_params()  # lmfit.Parameters()

    params["vg4_amplitude"].set(
        value=5,
        min=0,  # max=params["vg2_amplitude"].value
    )
    params["vg4_center"].set(value=136, min=135, max=137)
    params["vg4_sigma"].set(value=0.2, min=0.1, max=0.5)
    params["vg4_gamma"].set(value=0.2, max=0.5)

    # params["vg3_amplitude"].set(expr="vg4_amplitude * 0.75")
    params["vg3_center"].set(expr="vg4_center+4.4")
    params["vg3_sigma"].set(expr="vg4_sigma")
    params["vg3_gamma"].set(expr="vg4_gamma")

    params["vg2_amplitude"].set(
        value=5,  # The amplitude refers to the AREA of the peak, not height.
        min=0,
    )
    params["vg2_center"].set(
        value=137.5, min=136.5, max=138.5
    )  # min=607, max=610
    params["vg2_sigma"].set(value=0.3, min=0.1, max=0.4)
    params["vg2_gamma"].set(value=0.2, max=0.5)

    # params["vg1_amplitude"].set(expr="vg2_amplitude * 0.75")
    params["vg1_center"].set(expr="vg2_center+4.4")
    params["vg1_sigma"].set(expr="vg2_sigma")
    params["vg1_gamma"].set(expr="vg2_gamma")

    params.update(background_model.guess(data=y, x=x))

    # Linear section

    params["lin_intercept"].set(value=min(y), vary=False)
    # params["lin_slope"].set(value=slope)

    # Shirley section
    # params["shirley_const"].set(value=y[0], max=min(y))
    # params["shirley_k"].set(value=0.003)

    # Fit
    result = combined_model.fit(
        y,
        params,
        # method="basinhopping",
        x=x,
        y=y,
    )  

    # print(result.fit_report())
    fit_report_name = Path("fit_report_" + file.name)
    with open(fit_report_name, "w") as f:
        f.write(result.fit_report())
    fit_report_name.rename(fit_reports_directory / fit_report_name)

    # shirley = result.eval_components()["shirley_"]

    comps = result.eval_components()

    lin = result.eval_components()["lin_"]

    # matplotlib
    fig, ax = plt.subplots()

    ax.scatter(x, y, s=1, c="black", label="data")
    # plt.plot(x, result.init_fit, "--", label="initial fit")
    ax.plot(x, result.best_fit, "-", c="darkorange", label="best fit")  # -bg
    ax.plot(x, lin, ":", c="firebrick", label="linear bg", linewidth=1)
    # ax.plot(x, lin, ".-", label="linear bg")

    ax.plot(x, comps["vg1_"] + lin, "--", c="lightgreen", label="voigt1")
    ax.plot(x, comps["vg2_"] + lin, "--", c="lightblue", label="voigt2")
    ax.plot(x, comps["vg3_"] + lin, "--", c="darkgreen", label="voigt3")
    ax.plot(x, comps["vg4_"] + lin, "--", c="darkblue", label="voigt4")

    ax.set_title("CTF3 data with 4 fft voigt peak and linear background")
    ax.set_ylabel("Intensity [a.u.]")
    ax.set_xlabel("E_kin [eV]")
    ax.legend()

    fig.savefig(str(fit_plots_directory / file.name) + ".png", dpi=300)
    plt.close()

    return result

# Perform fit --------------------------------------------------------------------------------

start = time.time_ns()

file_path = Path(
    path/to/0-CTF3_21&31.csv"
)

for file in sorted(file_path.parent.resolve().iterdir()):
    if file.is_file():
        print(file.name)
        result = fit_four_fftvoigt_binding(
            file, 0, 0, 0, 0, 0, 0, 0, 0
        )
        save_modelresult(
            result,
            str(file_path.resolve().parent)
            + "/fit_reports/"
            + file.name
            + ".sav",
        )
        break
    else:
        print(file.name + " is directory - no fitting possible")
end = time.time_ns()
elapsed_time = end - start

print("Execution time:", elapsed_time, "nano seconds")


```

**Comment by SunC0der** [comment]
> Even if using the binding energy scale it is common to present the data in a way so that the kinetic energy is increaing from left to right. I would recommend to simply add ax.set_xlim(max(x), min(x)).
You mean just to change the display of the x-axis ? Like with ax.invert_xaxis() ? 

> The problem is probably caused by the peak currently not being limited to always have a positive amplitude. That's due to the norming to area in the model function. Changing the norm to abs(area) fixes this

Yes, it fixes it for me as well, thank you. 
But how could the area even get negative if both amplitudes of the lorentzian and the gaussian are set to +1.0 ?

**Comment by Julian-Hochhaus** [comment]
> But how could the area even get negative if both amplitudes of the lorentzian and the gaussian are set to +1.0 ?

The convolution is functioning properly, and the negative area observed is the expected outcome due to the decreasing values in your x-array. This means that the Simpson's rule is being calculated from right to left instead of the usual left to right. In XPS, it is established that the area should be counted positively. Hence, we need to be cautious by either reversing the order in x or considering the absolute value of the area :).
![image](https://github.com/Julian-Hochhaus/lmfit-additional-models/assets/22680429/702fdfec-48f4-4e66-9fd5-482c0f1c9f45)

It's a bit of a hidden quirk that can be hard to catch while coding, but once you think about it, it starts to make complete sense (... and I must admit that I stumbled upon this question myself in the past).

**Comment by SunC0der** [comment]
> This means that the Simpson's rule is being calculated from right to left instead of the usual left to right 
Are you sure ? In the official scipy documentation nothing is written about a changed order of integrating: 
https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.simpson.html 
The descending x-values, instead of the normally ascending x-values, are already enough to turn the integration results negative, right ? 
> Hence, we need to be cautious by either reversing the order in x or considering the absolute value of the area :).
Absolutely, now the abs() solution is applied in my code.
> It's a bit of a hidden quirk that can be hard to catch while coding, but once you think about it, it starts to make complete sense (... and I must admit that I stumbled upon this question myself in the past).
Feels good to know, I am still confused sometimes whether my math-, my coding- or my understanding-of-the-problem skills prevent me from debugging my scripts ...

**Comment by Julian-Hochhaus** [comment]
> Are you sure ? In the official scipy documentation nothing is written about a changed order of integrating:
> https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.simpson.html
> The descending x-values, instead of the normally ascending x-values, are already enough to turn the integration results negative, right ?

Yes, decreasing x values 'from left to right' area already enough to turn the area negative. I found it out by stumpling upon the same problem and already wondered, why it is not mentioned in the documentation. But the code speaks for itself. 
[If the values in x are decreasing, the dx turns negative. ](https://github.com/scipy/scipy/blob/c1ed5ece8ffbf05356a22a8106affcd11bd3aee0/scipy/integrate/_quadrature.py#L536-L664).

[In addition, there are some questions on stackoverflow, mentioning the same question](https://stackoverflow.com/questions/57996849/simpsons-rule-integration-negative-area)
